src/data_access/api/SpotifyAPI.java

```diff
@@ -0,0 +1,109 @@
+        this.CLIENT_ID = dotenv.get("CLIENT_ID");
+        this.CLIENT_SECRET = dotenv.get("CLIENT_SECRET");
+        this.client = new OkHttpClient().newBuilder().build();
+        this.songFactory = songFactory;
+    }
+
+    @Override
+    public Song getRandomSongFromGenre(String genre) {
+        JSONObject responseBody = retrieveSongData(genre);
+        JSONArray songsArray = responseBody.getJSONObject("tracks").getJSONArray("items");
+        JSONObject item;
+        Song song = null;
+        int i = 0;
+        do{
+            item = (JSONObject) songsArray.get(i);
+            if (item.get("preview_url") instanceof String){
+                JSONObject albumArtistInfo = (JSONObject) item.getJSONObject("album").getJSONArray("artists").get(0);
+                String songName = item.getString("name");
+                String audio = item.getString("preview_url");
+                String artistName = albumArtistInfo.getString("name");
+                song = songFactory.create(songName, artistName, audio);
+            }
+            i++;
+        }while (!(item.get("preview_url") instanceof String)  || item.getInt("popularity") < 83);
+
+        if(song == null){
+            return getRandomSongFromGenre(genre);
+        }else {
+            return song;
+        }
+    }
+    private JSONObject retrieveSongData(String genre)  {
+        String url = getRandomEndpoint(genre);
+        JSONObject response;
+        try {
+            Request request = new Request.Builder()
+                    .url(url)
+                    .addHeader("Authorization", "Bearer " + authToken)
+                    .addHeader("Content-Type", "application/json")
+                    .build();
+
+            Response responseRequest = client.newCall(request).execute();
+
+            if(responseRequest.code() == 401){ // Code 401 - bad (expired or null) auth token
+                generateAuthToken();
+                return retrieveSongData(genre); // recall api with auth token
+            }
+            if(responseRequest.body() == null) {throw new RuntimeException("Error - no response body");}
+            response = new JSONObject(responseRequest.body().string());
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        return response;
+    }
+    private void generateAuthToken(){
+        String endpoint = "https://accounts.spotify.com/api/token";
+        RequestBody body = new FormBody.Builder()
+                .add("client_id", CLIENT_ID)
+                .add("client_secret", CLIENT_SECRET)
+                .add("grant_type", "client_credentials")
+                .build();
+        Request request = new Request.Builder()
+                .url(endpoint)
+                .addHeader("Content-Type", "application/x-www-form-urlencoded")
+                .post(body)
+                .build();
+        try{
+            Response response = client.newCall(request).execute();
```

Comment:
```suggestion
        try (Response response = client.newCall(request).execute()) {
```
Same thing as above.